sway-core/src/language/ty/expression/expression.rs

```diff
@@ -429,4 +433,81 @@ impl TyExpression {
     pub(crate) fn extract_literal_value(&self) -> Option<Literal> {
         self.expression.extract_literal_value()
     }
+
+    // Checks if this expression references a deprecated item
+    pub(crate) fn check_deprecated(
+        &self,
+        engines: &Engines,
+        handler: &Handler,
+        allow_deprecated: &mut AllowDeprecatedState,
+    ) {
+        fn emit_warning_if_deprecated(
+            attributes: &AttributesMap,
+            span: &Span,
+            handler: &Handler,
+            message: &str,
+            allow_deprecated: &mut AllowDeprecatedState,
+        ) {
+            if allow_deprecated.is_allowed() {
+                return;
+            }
+
+            if let Some(v) = attributes
+                .get(&AttributeKind::Deprecated)
+                .and_then(|x| x.last())
+            {
+                let mut message = message.to_string();
+
+                if let Some(sway_ast::Literal::String(s)) = v
+                    .args
+                    .iter()
+                    .find(|x| x.name.as_str() == "note")
+                    .and_then(|x| x.value.as_ref())
+                {
+                    message.push_str(": ");
+                    message.push_str(s.parsed.as_str());
+                }
+
+                handler.emit_warn(CompileWarning {
+                    span: span.clone(),
+                    warning_content: Warning::UsingDeprecated { message },
+                })
+            }
+        }
+
+        match &self.expression {
+            TyExpressionVariant::StructExpression {
+                struct_ref,
+                instantiation_span,
+                ..
+            } => {
+                let s = engines.de().get(struct_ref.id());
+                emit_warning_if_deprecated(
+                    &s.attributes,
+                    instantiation_span,
+                    handler,
+                    "deprecated struct",
+                    allow_deprecated,
+                );
+            }
+            TyExpressionVariant::FunctionApplication {
+                call_path, fn_ref, ..
+            } => {
+                if let Some(TyDecl::ImplTrait(t)) = engines.de().get(fn_ref).implementing_type {
+                    let t = engines.de().get(&t.decl_id).implementing_for;
+                    if let TypeInfo::Struct(struct_ref) = engines.te().get(t.type_id) {
+                        let s = engines.de().get(struct_ref.id());
+                        emit_warning_if_deprecated(
+                            &s.attributes,
+                            &call_path.span(),
+                            handler,
+                            "deprecated struct",
+                            allow_deprecated,
+                        );
+                    }
+                }
+            }
+            _ => {}
```

Comment:
Perhaps just adding a TODO that we want to cover other deprecated items in the future.

Reply:
Done.